### PR TITLE
refactor: move IDisposable to core interfaces

### PR DESCRIPTION
**Pull request by RishhiB**
In preparation to remove some shared interfaces from common-definitions, I have copied the implementation of this interface from common-definitions to core-interfaces.

### Moved to core-interfaces

- interface IDisposable

Related to [AB#4057](https://dev.azure.com/fluidframework/235294da-091d-4c29-84fc-cdfc3d90890b/_workitems/edit/4057)

### CODE REVIEW

File: common/lib/common-definitions/CHANGELOG.md [hunk 1]
```diff
@@ -0,0 +1,12 @@
+# @fluidframework/common-definitions Changelog
+
+## [1.2.0](https://github.com/microsoft/FluidFramework/releases/tag/common-definitions_v1.2.0)
+
+### Deprecated interface
+
+The following interface is deprecated in this release. The implementations have been moved to other
+packages.
```
Comment (alexvy86) [inline common/lib/common-definitions/CHANGELOG.md:8]: Nit. Just to make it clear it's the interfaces themselves that are moving, not classes that implement the interfaces.
```suggestion
The following interface is deprecated in this release. Its replacement has been moved to another package.
```

**Review by alexvy86 — COMMENTED** [review]
One phrasing nit below. Also, for the packages where we had to add the new dependency on core-interfaces, did we check that we can't remove the one for common-definitions? They might be using other things from it so it's fine if we can't, but now would be a good time to make sure.

**Review by tylerbutler — COMMENTED** [review]
Looks good to me, but lets get sign off from some other folks before merging.

**Comment by markfields** [comment]
As we were chatting elsewhere about moving stuff to client packages, I looked at the `README` for `core-interfaces`, here's what it says:

> Core interfaces describing the contracts the Fluid Loader supports or consumes.
> See also: [The Fluid Loader README](https://github.com/RishhiB/FluidFramework-1/blob/move-IDisposable/packages/common/README.md).

If we're going to be putting general-purpose (beyond host-loader interfaces) in here, can we update the README to describe this broader scope?

cc @anthony-murphy

**Comment by anthony-murphy** [comment]
> As we were chatting elsewhere about moving stuff to client packages, I looked at the `README` for `core-interfaces`, here's what it says:
> 
> > Core interfaces describing the contracts the Fluid Loader supports or consumes.
> > See also: [The Fluid Loader README](https://github.com/RishhiB/FluidFramework-1/blob/move-IDisposable/packages/common/README.md).
> 
> If we're going to be putting general-purpose (beyond host-loader interfaces) in here, can we update the README to describe this broader scope?
> 
> cc @anthony-murphy

i think that comment is imprecise. core-interface doesn't container the interfaces the loader must implement, those are in container-definitions. these interfaces are all interfaces that cross the loader boundary from the host into the runtime, or runtime out to the host. Even that isn't totally true. i think its fine to have a utility type used across all our layers here.

**Comment by anthony-murphy** [comment]
> One phrasing nit below. Also, for the packages where we had to add the new dependency on core-interfaces, did we check that we can't remove the one for common-definitions? They might be using other things from it so it's fine if we can't, but now would be a good time to make sure.

i don't see anywhere were we added a new dependency. in all cases i see it was just moved from an implicit to an explicit dependency. our common (definition) interfaces form a DAG: core->drivers->container->runtime

**Review by anthony-murphy — APPROVED** [review]

**Comment by alexvy86** [comment]
> i don't see anywhere were we added a new dependency. in all cases i see it was just moved from an implicit to an explicit dependency. our common (definition) interfaces form a DAG: core->drivers->container->runtime

My point was that if a package previously depended on common-utils only for things that moved to core-utils and now its package.json has dependency entries for both, we could remove the entry for common-utils.

**Comment by DLehenbauer** [comment]
FYI - Regarding IDisposable, you may want to have a look at this:
https://www.totaltypescript.com/typescript-5-2-new-keyword-using

**Comment by RishhiB** [comment]
I was going to do Alejandro's point of removing unnecessary common-definitions as a dep after moving events and logging

**Pull request merged**